src/zspec/zcl/utils.ts

```diff
@@ -32,37 +31,73 @@ export function getDataTypeClass(dataType: DataType): DataTypeClass {
     let name: string;
+    // if manufacturer code is given, consider partial match if didn't match against manufacturer code
+    let partialMatch: boolean = (manufacturerCode != null);
 
-    const clusters: CustomClusters = {
-        ...customClusters, // Custom clusters have a higher priority than Zcl clusters (custom cluster has a DIFFERENT name than Zcl clusters)
-        ...Clusters,
-        ...customClusters, // Custom clusters should override Zcl clusters (custom cluster has the SAME name than Zcl clusters)
-    };
+    for (const clusterName in clusters) {
+        const cluster = clusters[clusterName as ClusterName];
 
-    if (typeof key === 'number') {
-        if (manufacturerCode) {
-            name = Object.entries(clusters)
-                .find((e) => e[1].ID === key && e[1].manufacturerCode === manufacturerCode)?.[0];
-        } 
-        
-        if (!name) {
-            name = Object.entries(clusters).find((e) => e[1].ID === key && !e[1].manufacturerCode)?.[0];
+        if (cluster.ID === id) {
+            // priority on first match when matching only ID
+            /* istanbul ignore else */
+            if (name == undefined) {
+                name = clusterName;
+            }
+
+            if (manufacturerCode && (cluster.manufacturerCode === manufacturerCode)) {
+                name = clusterName;
+                partialMatch = false;
+                break;
+            } else if (!cluster.manufacturerCode) {
+                name = clusterName;
+                break;
+            }
         }
+    }
+
+    return [name, partialMatch];
+}
+
+function getClusterDefinition(key: string | number, manufacturerCode: number | null, customClusters: CustomClusters)
+    : {name: string, cluster: ClusterDefinition} {
+    let name: string;
+
+    if (typeof key === 'number') {
+        let partialMatch: boolean;
+
+        // custom clusters have priority over Zcl clusters, except in case of better match (see below)
+        [name, partialMatch] = findClusterNameByID(key, manufacturerCode, customClusters);
 
         if (!name) {
-            name = Object.entries(clusters).find((e) => e[1].ID === key)?.[0];
+            [name, partialMatch] = findClusterNameByID(key, manufacturerCode, Clusters);
+        } else if (partialMatch) {
+            let zclName: string;
+            [zclName, partialMatch] = findClusterNameByID(key, manufacturerCode, Clusters);
+
+            // Zcl clusters contain a better match, use that one
+            if (zclName != undefined && !partialMatch) {
+                name = zclName;
+            }
         }
     } else {
         name = key;
     }
 
-    let cluster = clusters[name];
+    let cluster = (name != undefined) && hasCustomClusters(customClusters) ?
+        {
+            ...Clusters[name as ClusterName],
+            ...customClusters[name],// should override Zcl clusters
```

Comment:
I think this old logic is missing here:

```
        ...customClusters, // Custom clusters have a higher priority than Zcl clusters (custom cluster has a DIFFERENT name than Zcl clusters)
        ...Clusters,
        ...customClusters, // Custom clusters should override Zcl clusters (custom cluster has the SAME name than Zcl clusters)
```

Besides that, this commit looks OK

Reply:
The same logic should be applied in the order the lookup is now done (looks into custom clusters first).
Current tests are passing, but if you want to add a couple very specific in the way you intend this to work?

Reply:
It's not tested by tests yet, but it would happen if you define a cluster with an already existing ID.

Reply:
Isn't this one doing it?
https://github.com/Koenkk/zigbee-herdsman/blob/f1890adb3678c86f209122b0cc7ddb6f7eb8a497/test/controller.test.ts#L2438

Reply:
This tests defines a attribute for an existing cluster, what I mean is defining a new cluster of which the ID is already used by an existing cluster.

Reply:
This test is the same as the one linked above, except it should override the ZCL cluster. (`name` is not ZCL, but ID is.)
Can you confirm that is correct?
```typescript
    it('Should allow to specific custom cluster as override for Zcl cluster', async () => {
        await controller.start();
        await mockAdapterEvents['deviceJoined']({networkAddress: 129, ieeeAddr: '0x129'});
        const device = controller.getDeviceByIeeeAddr('0x129');
        device.addCustomCluster('myCustomCluster', {ID: Zcl.Clusters.genBasic.ID, commands: {}, commandsResponse: {}, attributes: {customAttr: {ID: 256, type: Zcl.DataType.UINT8}}});
        const buffer = Buffer.from([24,169,10,0,1,24,3,0,0,24,1]);
        const header = Zcl.Header.fromBuffer(buffer);
        await mockAdapterEvents['zclPayload']({wasBroadcast: false, address: 129, clusterID: Zcl.Clusters.genBasic.ID, data: buffer, header, endpoint: 1, linkquality: 50, groupID: 1});
        expect(events.message.length).toBe(1);
        expect(events.message[0].data).toStrictEqual({customAttr: 3, 0: 1/*zclVersion no longer recognized, cluster is overridden*/});
        expect(events.message[0].cluster).toBe('myCustomCluster');
    });
```

Reply:
I updated the test in https://github.com/Koenkk/zigbee-herdsman/pull/1042/commits/a97271bd1a798d9775bde591f8ffe9a1d1195d3b and it passes, looks good!

Reply:
That's just attribute overrides in your commit though. Should I add the test mentioned above for complete cluster override too ("same cluster ID, different name", results in custom cluster used)?

Reply:
Yes please!